src/WindowsAzureMessaging/WindowsAzureMessaging/Models/MSInstallationTemplate.m

```diff
@@ -0,0 +1,133 @@
+//----------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//----------------------------------------------------------------
+
+#import "MSInstallationTemplate.h"
+#import "MSTagHelper.h"
+
+@interface MSInstallationTemplate ()
+
+@property(nonatomic, copy) NSSet<NSString *> *tags;
+@property(nonatomic, copy) NSDictionary<NSString *, NSString *> *headers;
+
+@end
+
+@implementation MSInstallationTemplate
+
+@synthesize body;
+@synthesize tags;
+@synthesize headers;
+
+- (instancetype)init {
+    if (self = [super init]) {
+        tags = [NSSet new];
+        headers = [NSDictionary new];
+    }
+    return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)coder {
+    if (self = [super init]) {
+        body = [coder decodeObjectForKey:@"body"] ?: @"";
+        tags = [coder decodeObjectForKey:@"tags"];
+        headers = [coder decodeObjectForKey:@"headers"];
+    }
+
+    return self;
+}
+
+#pragma mark Tags
+
+- (BOOL)addTag:(NSString *)tag {
+    return [self addTags:[NSArray arrayWithObject:tag]];
+}
+
+- (BOOL)addTags:(NSArray<NSString *> *)tagsToAdd {
+    NSMutableSet *tmpTags = [NSMutableSet setWithSet:self.tags];
+
+    for (NSString *tag in tagsToAdd) {
+        if (isValidTag(tag)) {
+            [tmpTags addObject:tag];
+        } else {
+            NSLog(@"Invalid tag: %@", tag);
+            return NO;
+        }
+    }
+
+    self.tags = [tmpTags copy];
+    return YES;
+}
+
+- (BOOL)removeTag:(NSString *)tag {
+    return [self removeTags:[NSArray arrayWithObject:tag]];
+}
+
+- (BOOL)removeTags:(NSArray<NSString *> *)tagsToRemove {
+    NSMutableSet *tmpTags = [NSMutableSet setWithSet:self.tags];
+
+    [tmpTags minusSet:[NSSet setWithArray:tagsToRemove]];
```

Comment:
Using `minusSet` doesn't allow us to track if the set was modified. If we are going to return `BOOL`, then we should try to make it accurate. That probably means we need to iterate over the tags to remove, check if the tag exists, remove it, and mark the set as "modified".

Reply:
Adding change tracking via `isDirty` and KVO